File: algorithm/java/AlphabetBoardPath.java
import java.io.*;
import java.util.*;
import com.eclipsesource.json.JsonArray;


class Solution {
    public String alphabetBoardPath(String target) {
        int x = 0, y = 0;
        StringBuilder sb = new StringBuilder();
        for (char ch : target.toCharArray()) {
            int x1 = (ch - 'a') % 5, y1 = (ch - 'a') / 5;
            sb.append(
                    String.join("", Collections.nCopies(Math.max(0, y - y1), "U"))
                            + String.join("", Collections.nCopies(Math.max(0, x1 - x), "R"))
                            + String.join("", Collections.nCopies(Math.max(0, x - x1), "L"))
                            + String.join("", Collections.nCopies(Math.max(0, y1 - y), "D"))
                            + "!");
            x = x1;
            y = y1;
        }
        return sb.toString();
    }
}

public class AlphabetBoardPath {
    public static String stringToString(String input) {
        return JsonArray.readFrom("[" + input + "]").get(0).asString();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = in.readLine()) != null) {
            String target = line;
            String ret = new Solution().alphabetBoardPath(target);
            String out = (ret);
            System.out.print(out + "\n");
            break;
        }
    }
}

File: algorithm/python/alphabet_board_path.py
class Solution:
    def alphabetBoardPath(self, target):
        """
        Time Complexity: O(N)
        Space Complexity: O(N)
        """
        m = {c: [i // 5, i % 5] for i, c in enumerate("abcdefghijklmnopqrstuvwxyz")}
        x0, y0 = 0, 0
        res = []
        for c in target:
            x, y = m[c]
            if y < y0:
                res.append("L" * (y0 - y))
            if x < x0:
                res.append("U" * (x0 - x))
            if x > x0:
                res.append("D" * (x - x0))
            if y > y0:
                res.append("R" * (y - y0))
            res.append("!")
            x0, y0 = x, y
        return "".join(res)


def stringToString(input):
    import json

    return json.loads(input)


def main():
    import sys
    import io

    def readlines():
        for line in io.TextIOWrapper(sys.stdin.buffer, encoding="utf-8"):
            yield line.strip("\n")

    lines = readlines()
    while True:
        try:
            line = next(lines)
            #target = stringToString(line)
            target = line
            ret = Solution().alphabetBoardPath(target)
            out = ret
            print(out)
            break
        except StopIteration:
            break


if __name__ == "__main__":
    main()
